homeassistant/components/transmission/__init__.py

```diff
@@ -55,97 +46,182 @@
                 vol.Optional(CONF_USERNAME): cv.string,
                 vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
                 vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-                vol.Optional(TURTLE_MODE, default=False): cv.boolean,
                 vol.Optional(
                     CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
-                ): cv.time_period,
-                vol.Optional(
-                    CONF_MONITORED_CONDITIONS, default=["current_status"]
-                ): vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+                ): cv.positive_int,
             }
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
+async def async_setup(hass, config):
+    """Import the Transmission Component from config."""
+    if not hass.config_entries.async_entries(DOMAIN) and DOMAIN in config:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
     """Set up the Transmission Component."""
-    host = config[DOMAIN][CONF_HOST]
-    username = config[DOMAIN].get(CONF_USERNAME)
-    password = config[DOMAIN].get(CONF_PASSWORD)
-    port = config[DOMAIN][CONF_PORT]
-    scan_interval = config[DOMAIN][CONF_SCAN_INTERVAL]
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+
+    if not config_entry.options:
+        await async_populate_options(hass, config_entry)
+
+    client = TransmissionClient(hass, config_entry)
+    client_id = config_entry.entry_id
+    hass.data[DOMAIN][client_id] = client
+    if not await client.async_setup():
+        raise ConfigEntryNotReady
+
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload Transmission Entry from config_entry."""
+    hass.services.async_remove(DOMAIN, SERVICE_ADD_TORRENT)
+    if hass.data[DOMAIN][entry.entry_id].unsub_timer:
+        hass.data[DOMAIN][entry.entry_id].unsub_timer()
+
+    for component in "sensor", "switch":
+        await hass.config_entries.async_forward_entry_unload(entry, component)
+
+    del hass.data[DOMAIN]
+
+    return True
 
-    import transmissionrpc
-    from transmissionrpc.error import TransmissionError
 
+def get_api(host, port, username=None, password=None):
```

Comment:
Since we need to call this function from coroutines it seems better to make this a coroutine too. Then we can await the task on the executor thread pool inside this function instead.